src/connectors/impls/gpubsub/producer.rs

```diff
@@ -0,0 +1,298 @@
+    pub request_timeout: u64,
+    #[serde(default = "crate::connectors::impls::gpubsub::default_endpoint")]
+    pub endpoint: String,
+    pub topic: String,
+    #[cfg(test)]
+    #[serde(default = "crate::connectors::impls::gpubsub::default_skip_authentication")]
+    pub skip_authentication: bool,
+}
+
+impl ConfigImpl for Config {}
+
+#[derive(Default, Debug)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait()]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        ConnectorType("gpubsub_producer".to_string())
+    }
+
+    async fn build(&self, alias: &str, config: &ConnectorConfig) -> Result<Box<dyn Connector>> {
+        if let Some(raw_config) = &config.config {
+            let config = Config::new(raw_config)?;
+
+            Ok(Box::new(GpubConnector { config }))
+        } else {
+            Err(ErrorKind::MissingConfiguration(alias.to_string()).into())
+        }
+    }
+}
+
+struct GpubConnector {
+    config: Config,
+}
+
+#[async_trait::async_trait()]
+impl Connector for GpubConnector {
+    async fn create_sink(
+        &mut self,
+        sink_context: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        let url = Url::parse(&self.config.endpoint)?;
+        let sink = GpubSink {
+            config: self.config.clone(),
+            url: url.clone(),
+            hostname: url
+                .host_str()
+                .ok_or_else(|| {
+                    ErrorKind::InvalidConfiguration(
+                        "gpubsub-publisher".to_string(),
+                        "Missing hostname".to_string(),
+                    )
+                })?
+                .to_string(),
+            client: None,
+        };
+        builder.spawn(sink, sink_context).map(Some)
+    }
+
+    async fn connect(&mut self, _ctx: &ConnectorContext, _attempt: &Attempt) -> Result<bool> {
+        Ok(true)
+    }
+
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Required
+    }
+}
+
+struct GpubSink {
+    config: Config,
+    url: Url<HttpsDefaults>,
+    hostname: String,
+
+    client: Option<PublisherClient<InterceptedService<Channel, AuthInterceptor>>>,
+}
+
+#[async_trait::async_trait()]
+impl Sink for GpubSink {
+    async fn connect(&mut self, _ctx: &SinkContext, _attempt: &Attempt) -> Result<bool> {
+        let mut channel = Channel::from_shared(self.config.endpoint.clone())?
+            .connect_timeout(Duration::from_nanos(self.config.connect_timeout));
+        if self.url.scheme() == "https" {
+            let tls_config = ClientTlsConfig::new()
+                .ca_certificate(Certificate::from_pem(googapis::CERTIFICATES))
+                .domain_name(self.hostname.clone());
+
+            channel = channel.tls_config(tls_config)?;
+        }
+
+        let channel = channel.connect().await?;
+        #[allow(unused_assignments, unused_mut)]
```

Comment:
I really find this super confusing to read, could we re-structure this part to, for example, have a create_publisher() function that has one implementation in `test` and one in `not(test)` to make it clear what code is executed when?